sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala

```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
   }
 }
 
+trait GpuNumberToTimestampUnaryExpression extends GpuUnaryExpression {
+  
+  override def dataType: DataType = TimestampType
+  override def outputTypeOverride: DType = DType.TIMESTAMP_MICROSECONDS
+
+  /**
+   * Test whether if input * multiplier will cause Long-overflow. In Math.multiplyExact, 
+   * if there is an integer-overflow, then it will throw an ArithmeticException "long overflow"
+   */
+  def checkLongMultiplicationOverflow(input: ColumnVector, multiplier: Long): Unit = {
+      withResource(input.max()) { maxValue =>
+        Math.multiplyExact(maxValue.getLong, multiplier)
+      }
+      withResource(input.min()) { minValue =>
+        Math.multiplyExact(minValue.getLong, multiplier)
+      }
+  }
+}
+
+case class GpuSecondsToTimestamp(child: Expression) extends GpuNumberToTimestampUnaryExpression {
+
+  private lazy val convertTo: GpuColumnVector => ColumnVector = child.dataType match {
+    case LongType =>
+      (input: GpuColumnVector) => {
+        checkLongMultiplicationOverflow(input.getBase, DateTimeConstants.MICROS_PER_SECOND)
+        val mul = withResource(Scalar.fromLong(DateTimeConstants.MICROS_PER_SECOND)) { scalar =>
+          input.getBase.mul(scalar)
+        }
+        withResource(mul) { mul =>
+          mul.asTimestampMicroseconds()
+        }
+      }
+    case DoubleType | FloatType =>
+      (input: GpuColumnVector) => {
+        // basicly copied from GpuCast
+        withResource(Scalar.fromLong(DateTimeConstants.MICROS_PER_SECOND)) { microsPerSec =>
+          withResource(input.getBase.nansToNulls()) { inputWithNansToNull =>
+            withResource(FloatUtils.infinityToNulls(inputWithNansToNull)) {
+              inputWithoutNanAndInfinity =>
+                if (child.dataType == FloatType) {
+                  withResource(inputWithoutNanAndInfinity.castTo(DType.FLOAT64)) { doubles =>
+                    withResource(doubles.mul(microsPerSec, DType.INT64)) {
+                      inputTimesMicrosCv =>
+                        inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                    }
+                  }
+                } else {
+                  withResource(inputWithoutNanAndInfinity.mul(microsPerSec, DType.INT64)) {
+                    inputTimesMicrosCv =>
+                      inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                  }
+                }
+            }
+          }
+        }
+      }
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        println("DecimalType " + dt.precision + " " + dt.scale)
+        val decimal128Type = DType.create(DType.DTypeEnum.DECIMAL128, -6)
```

Comment:
Why -6 is used for the scale here ?

Reply:
SecondsToTimestamp does support decimals with a scale of 6 or less, which can be represented as microseconds. An exception will be thrown if the scale is more than 6. I updated the comment here.